roomserver/internal/alias.go

```diff
@@ -183,6 +186,65 @@ func (r *RoomserverInternalAPI) RemoveRoomAlias(
 		}
 	}
 
+	ev, err := r.DB.GetStateEvent(ctx, roomID, gomatrixserverlib.MRoomCanonicalAlias, "")
+	if err != nil && err != sql.ErrNoRows {
+		return err
+	}
+
+	stateAlias := gjson.GetBytes(ev.Content(), "alias").Str
+	// the alias to remove is currently set as the canonical alias, remove it
+	if stateAlias == request.Alias {
+		res, err := sjson.DeleteBytes(ev.Content(), "alias")
+		if err != nil {
+			return err
+		}
+
+		stateRes := &api.QueryLatestEventsAndStateResponse{}
+
+		if err = r.QueryLatestEventsAndState(ctx, &api.QueryLatestEventsAndStateRequest{RoomID: roomID}, stateRes); err != nil {
+			return err
+		}
+
+		stateEventIDs := make([]string, len(stateRes.StateEvents))
+		for i, ev := range stateRes.StateEvents {
+			stateEventIDs[i] = ev.EventID()
+		}
+
+		sender := request.UserID
+		if request.UserID != ev.Sender() {
+			sender = ev.Sender()
+		}
+		builder := &gomatrixserverlib.EventBuilder{
+			Sender:     sender,
+			RoomID:     ev.RoomID(),
+			Type:       ev.Type(),
+			StateKey:   ev.StateKey(),
+			PrevEvents: []string{ev.EventID()},
+			AuthEvents: stateEventIDs,
+			Content:    res,
+		}
+
+		newEvent, err := builder.Build(time.Now(), r.ServerName, r.Cfg.Matrix.KeyID, r.Cfg.Matrix.PrivateKey, ev.RoomVersion)
+		if err != nil {
+			return err
+		}
+
+		inputEvent := api.InputRoomEvent{
+			Kind:          api.KindNew,
+			Event:         newEvent.Headered(ev.RoomVersion),
+			Origin:        r.ServerName,
+			StateEventIDs: stateEventIDs,
```

Comment:
It's best not to specify `StateEventIDs` and `HasState` here — we only use those if we want to force the roomserver to create a new state snapshot, which isn't necessary in this case.

This means we should be able to remove all of the `r.QueryLatestEventsAndState` bit.